Decorator fixes

andreibratu opened the pull request:
* Use /otel endpoint instead of processing spans locally
* Helpful error messages on evaluations.run()
* Throw errors for invalid decorator uses
* Decorators will no longer serialize all errors but raise critical ones -> evaluations will not hang mysteriously
* Move away from context variables in favor of OTEL's runtime context API
* @flow will also pick up SDK logging e.g. prompts.call(...) or tool.log(...)
* Sync with backend API: jinja release, log status use

Commit:
Refactoring code

Commit:
simplify decorators

Commit:
Use /otel endpoint to process endpoints

andreibratu commented:
Test script
```python
from humanloop import Humanloop
from openai import OpenAI

hl_client = Humanloop(
    api_key="<HL_API_KEY>",
    base_url="http://localhost:80/v5",
)

TEMPLATE = "You are a useful chatbot that speaks like a {{personality}}"

client = OpenAI(api_key="<OPENAI_KEY>")
hl_client.prompts.populate_template(TEMPLATE, inputs={"personality": "pirate"})
messages = [
    {
        "role": "system",
        "content": "You are a useful chatbot that speaks like a ",
    }
]


@hl_client.tool(path="Andrei QA/Calculator")
def calculator(a: int, b: int):
    return a + b


@hl_client.flow(path="Andrei QA/Flow")
def fn(cool: str, beans: int):
    output = calculator(1, 2)
    hl_client.tools.log(
        path="Andrei QA/Log",
        tool={
            "function": {
                "name": "calculator",
                "description": "Adds two numbers",
                "parameters": {
                    "a": {
                        "type": "int",
                        "description": "First number",
                    },
                    "b": {
                        "type": "int",
                        "description": "Second number",
                    },
                },
            }
        },
        inputs={
            "a": 1,
            "b": 2,
        },
        output=output,
    )

    hl_client.prompts.call(
        path="Andrei QA/Call Prompt",
        prompt={
            "provider": "openai",
            "model": "gpt-4o",
            "temperature": 0.8,
            "frequency_penalty": 0.6,
            "presence_penalty": 0.6,
        },
        messages=[
            {
                "content": "Say something funny",
                "role": "system",
            }
        ],
    )

    with hl_client.prompt(path="Andrei QA/Prompt"):
        while True:
            user_input = input("> ")
            if user_input == "exit":
                break
            messages.append(
                {
                    "role": "user",
                    "content": user_input,
                }
            )
            response = (
                client.chat.completions.create(
                    model="gpt-4o",
                    messages=messages,
                    temperature=0.8,
                    frequency_penalty=0.6,
                    presence_penalty=0.6,
                    seed=42,
                )
                .choices[0]
                .message.content
            )
            print(response)
            messages.append(
                {
                    "role": "assistant",
                    "content": response,
                }
            )

    return messages[-1]


fn("cool", 1)

# @hl_client.flow(path="Andrei QA/Flow Evaluate")
# def fn_evaluate(a: int, b: int):
#     return calculator(a, b)


# hl_client.evaluations.run(
#     file={
#         "path": "Andrei QA/Flow Evaluate",
#         "callable": fn_evaluate,
#     },
#     name="Test",
#     dataset={
#         "path": "Andrei QA/Arithmetic",
#         "datapoints": [
#             {
#                 "inputs": {
#                     "a": 1,
#                     "b": 2,
#                 },
#                 "target": {
#                     "value": 3,
#                 },
#             },
#             {
#                 "inputs": {
#                     "a": 3,
#                     "b": 4,
#                 },
#                 "target": {
#                     "value": 7,
#                 },
#             },
#         ],
#     },
#     evaluators=[
#         {
#             "path": "Andrei QA/Equals",
#             "callable": lambda x, y: x["output"] == y["target"]["value"],
#             "return_type": "boolean",
#             "args_type": "target_required",
#         }
#     ],
# )
```

Commit:
changed name of prompt context

jamesbaskerville commented on src/humanloop/client.py:
```diff
@@ -136,22 +135,20 @@ def __init__(
         *,
-        path: Optional[str] = None,
-        **prompt_kernel: Unpack[DecoratorPromptKernelRequestParams],  # type: ignore
+        path: str,
```
nit for all of these: we should support passing either path or id, no?

andreibratu replied:
I'd keep it to path to reduce cognitive load

jamesbaskerville commented on src/humanloop/client.py:
```diff
@@ -136,22 +135,20 @@ def __init__(
         else:
             self._opentelemetry_tracer = opentelemetry_tracer
 
+    @contextmanager
```
as discussed earlier, this is cool but probably unnecessary -- simpler to just keep it as a decorator on a callable

andreibratu replied:
yep done

jamesbaskerville commented on src/humanloop/otel/exporter.py:
```diff
@@ -128,265 +110,64 @@ def force_flush(self, timeout_millis: int = 10000) -> bool:
         return True
 
     def _do_work(self):
-        """Upload spans to Humanloop.
-
-        Ran by worker threads. The threads use the self._shutdown flag to wait
-        for Spans to arrive. Setting a timeout on self._upload_queue.get() risks
-        shutting down the thread early as no Spans are produced e.g. while waiting
-        for user input into the instrumented feature or application.
-
-        Each thread will upload a Span to Humanloop, provided the Span has all its
-        dependencies uploaded. The dependency happens in a Flow Trace context, where
-        the Trace parent must be uploaded first. The Span Processor will send in Spans
-        bottoms-up, while the upload of a Trace happens top-down. If a Span did not
-        have its span uploaded yet, it will be re-queued to be uploaded later.
-        """
```
should still replace this with a correct docstring

andreibratu replied:
outdated

Commit:
refactoring

Commit:
wip

Commit:
QA refactor sdk with evals

Commit:
Error handling in decorators

Commit:
Merge branch 'master' into decorators-fixes

Commit:
update lock file

peadaroh commented:
Just to confirm this refactors the 'polling behaviour' for flow logs to instead check for when it is updated.

Commit:
QA pass on python

Commit:
Docstrings, small refactors

Commit:
mypy nit

Commit:
checking dependencies

jamesbaskerville commented on src/humanloop/client.py:
```diff
@@ -94,16 +98,14 @@ def __init__(
-        telemetry system. Otherwise, an internal TracerProvider will be used.
-        If you provide only the `TraceProvider`, the SDK will log under a Tracer
-        named `humanloop.sdk`.
+        The Humanloop SDK File decorators use OpenTelemetry internally.
+        You can provide a TracerProvider and a Tracer to integrate
+        with your existing telemetry system. If not provided,
+        an internal TracerProvider will be used.
```
Does our OTEL exporting stuff still work ok if user passes in a TracerProvider?

jamesbaskerville commented on src/humanloop/client.py:
````diff
@@ -176,89 +167,87 @@ def call_llm(messages):
+        If a different model, endpoint, or hyperparameter is used, a new
+        Prompt version is created. For example:
         ```
+        @humanloop_client.prompt(path="My Prompt")
````
nit: inconsistent, above example just uses `@prompt`. Should standardize

jamesbaskerville commented on src/humanloop/client.py:
```diff
@@ -176,89 +167,87 @@ def call_llm(messages):
             provided, the function name is used as the path and the File
             is created in the root of your Humanloop organization workspace.
 
         :param prompt_kernel: Attributes that define the Prompt. See `class:DecoratorPromptKernelRequestParams`
```
looks like we're getting rid of this?

jamesbaskerville commented on tests/integration/chat_agent/__init__.py:
are these tests all gone now? Or did they not work?

jamesbaskerville commented on src/humanloop/overload.py:
```diff
@@ -0,0 +1,124 @@
+    ]:
+        trace_id = get_trace_id()
+        if trace_id is not None and type(client) is FlowsClient:
+            context = get_decorator_context()
+            if context is None:
+                raise HumanloopRuntimeError("Internal error: trace_id context is set outside a decorator context.")
```
if I'm understanding error right, this should also be a check in the line 59 block? i.e. apply even when not flow client.

I do like all this validation/error checking though!

jamesbaskerville commented on src/humanloop/overload.py:
```diff
@@ -0,0 +1,124 @@
+            try:
+                response = self._log(**kwargs_eval)
+            except Exception as e:
+                # Re-raising as HumanloopDecoratorError so the decorators don't catch it
```
nice

jamesbaskerville commented on src/humanloop/otel/processor.py:
```diff
@@ -1,252 +1,54 @@
-    complete: bool
-
-
 class HumanloopSpanProcessor(SimpleSpanProcessor):
```
new docstring pls 🙏🏼

jamesbaskerville replied:
the info about the instrumentors is still relevant, no?

jamesbaskerville commented on src/humanloop/otel/processor.py:
```diff
@@ -1,252 +1,54 @@
+        if is_llm_provider_call(span):
+            decorator_context = get_decorator_context()
+            if decorator_context and decorator_context.type == "prompt":
+                path, template = (
+                    decorator_context.path,
+                    decorator_context.version["template"],
+                )
```
isn't path the only arg on prompt decorators now? or did I read that wrong and templates are still valid?

jamesbaskerville commented on src/humanloop/otel/processor.py:
```diff
@@ -1,252 +1,54 @@
+                return
+            evaluation_context = get_evaluation_context()
+            if evaluation_context is not None and evaluation_context.path == decorator_context.path:
+                # User made a provider call inside a @prompt context, ignore the span
```
fix: inside an evaluation context

also, is `evaluation_context.path` the prompt path being evaluated?

jamesbaskerville commented on src/humanloop/otel/processor.py:
```diff
@@ -1,252 +1,54 @@
+            if evaluation_context is not None and evaluation_context.path == decorator_context.path:
+                # User made a provider call inside a @prompt context, ignore the span
+                return
         self.span_exporter.export([span])
```
😍

jamesbaskerville replied:
the amount of redlines in this diff is amazing

jamesbaskerville commented on src/humanloop/decorators/prompt.py:
```diff
@@ -0,0 +1,44 @@
+                    type="prompt",
+                    version={
+                        # TODO: Implement a reverse-lookup of the template
+                        "template": None,
```
btw is this a breaking change? Ok that template can't be specified in the prompt decorator (I think it is -- duplicate with prompts.call right?)

jamesbaskerville commented on src/humanloop/decorators/tool.py:
```diff
@@ -31,75 +32,90 @@
+            evaluation_context = get_evaluation_context()
+            if evaluation_context is not None:
+                if evaluation_context.path == path:
+                    raise HumanloopRuntimeError("Tools cannot be evaluated with the `evaluations.run()` utility.")
```
Is this a TODO to support evaluating tools here?

jamesbaskerville commented on src/humanloop/decorators/tool.py:
```diff
@@ -31,75 +32,90 @@
-                        value=enhanced_tool_kernel,
-                    )
+                write_to_opentelemetry_span(
+                    span=span,  # type: ignore [arg-type]
```
why was this a new type issue?

jamesbaskerville commented on src/humanloop/decorators/tool.py:
```diff
@@ -31,75 +32,90 @@
+                    key=HUMANLOOP_FILE_KEY,
+                    value=tool_kernel,  # type: ignore [arg-type]
+                )
+                span.set_attribute(HUMANLOOP_FILE_PATH_KEY, path)
```
for myself -- equivalent to:
`write_to_opentelemetry_span(span=span, key=HUMANLOOP_FILE_PATH_KEY, value=path)`? And the helper just lets you easily write objects?

jamesbaskerville commented on src/humanloop/decorators/flow.py:
```diff
@@ -0,0 +1,133 @@
+                            "output_message": log_output_message,
+                            "id": this_flow_log.id,
+                        }
+                        # Write the Flow Log to the Span on HL_LOG_OT_KEY
```
typo: HUMANLOOP_LOG_KEY

jamesbaskerville commented on src/humanloop/evals/run.py:
```diff
@@ -245,39 +125,63 @@ def run_eval(
-    )
+    try:
+        hl_file = _upsert_file(file=file_, type=type_, client=client)
+    except ValidationError as e:
+        sys.stdout.write(f"{RED}Error in your `file` argument:\n\n{e}{RESET}")
+        return []
+    except Exception as e:
+        sys.stdout.write(f"{RED}Error in your `file` argument:\n\n{e}{RESET}")
+        return []
```
nit: you can combine these like `except (ValidationError, Exception) as e`. Actually could just be Exception since you're not handling them differently

jamesbaskerville commented on src/humanloop/evals/run.py:
```diff
@@ -315,61 +219,82 @@ def upload_callback(log_id: str):
+            try:
+                future.result()
+            except Exception as e:
+                sys.stderr.write(f"\n{RED}Error processing datapoint:\n{_get_error_message(e)}{RESET}\n")
+                _cancel_evaluation()
```
is it correct to cancel the whole eval when a single datapoint fails? seems excessive if I'm reading this right

jamesbaskerville commented on src/humanloop/evals/run.py:
```diff
@@ -467,10 +392,11 @@ def _wait_for_evaluation_to_complete(
             # Move the cursor up in stderr a number of lines equal to the number of lines in stats.progress
-            sys.stderr.write("\033[A" * (stats.progress.count("\n")))
+            # sys.stderr.write("\033[A" * (stats.progress.count("\n")))
             time.sleep(5)
```
fix/followup: we did say to move away from this polling behavior. Prefer separate PR for it though

jamesbaskerville commented on src/humanloop/evals/run.py:
```diff
@@ -521,48 +447,63 @@ def _get_checks(
+                "Using the `version` in `file` argument instead.\n"
+                f"{RESET}"
+            )
+            # TODO: document this
```
question: what does this mean?

jamesbaskerville commented on src/humanloop/evals/run.py:
```diff
@@ -893,6 +827,20 @@ def _check_evaluation_improvement(
+
+    # Extract the last 200 characters of the traceback
+    last_trace_part = (
+        (trace_info[-200:] + "..." if len(trace_info) > 200 else trace_info) if length_limit else trace_info
+    )
```
`...` should be in front if it's the last 200 chars, right?

jamesbaskerville replied:
also feels like `length_limit` should just be an int instead of a boolean

jamesbaskerville commented on src/humanloop/otel/exporter/__init__.py:
```diff
@@ -0,0 +1,153 @@
+                },
+                data=serialize_span(span_to_export),
+            )
+            print("RECV", span_to_export.attributes, response.json(), response.status_code)
```
fix: change to logger/debug

jamesbaskerville commented on src/humanloop/otel/exporter/__init__.py:
```diff
@@ -0,0 +1,153 @@
+                pass
+            else:
+                if eval_context_callback:
+                    print("HELLO")
```
remove

jamesbaskerville commented on src/humanloop/otel/exporter/proto.py:
```diff
@@ -0,0 +1,76 @@
+        ]
+    )
+
+    return MessageToJson(payload)
```
hmm not 100% sure this will work as expected -- OTLP JSON encoding differs a bit from the proto3 standard mapping. Possible we should just be working in proto anyways? Might make things simpler since it's more of the default rather than JSON.

https://opentelemetry.io/docs/specs/otlp/#json-protobuf-encoding

jamesbaskerville replied:
I guess it's fine if our IDs aren't "correct" as long as they're internally consistent?

jamesbaskerville commented on src/humanloop/otel/exporter/proto.py:
```diff
@@ -0,0 +1,76 @@
+                                attributes=[
+                                    KeyValue(
+                                        key=key,
+                                        value=AnyValue(string_value=str(value)),
```
are they always strings? Aren't they sometimes ints or some other type?

jamesbaskerville reviewed (commented):
generally looks great. Love the error/logging improvements and how simplified the processor/exporter logic is.

Few higher-level things:
- why are there more # type: ignore[arg-type] littered all over? what changed to cause that?
- need to handle the polling fix as a follow-on, though good we're sleeping in between polls now

jamesbaskerville committed:
fix: Add sleep to OTEL exporters (#51)

Infinite polling the export queue without a sleep was causing threads to take up a full core of CPU. Simple sleep fixes that.
